### PR TITLE
ARSN-406: add APIs in the needQuotas checks list

### PR DESCRIPTION
**Pull request by ghost**
Necessary changes to support quotas in S3, and cover the missing APIs to consider when evaluating quotas.

### CODE REVIEW

**Review by benzekrimaha — APPROVED** [review]

File: lib/policyEvaluator/RequestContext.ts [hunk 1]
```diff
@@ -15,9 +15,14 @@ import {
...
 
-const _actionNeedQuotaCheck = {
+export const actionNeedQuotaCheck = {
     objectPut: true,
+    objectPutVersion: true,
     objectPutPart: true,
+    objectRestore: true,
+    objectDelete: true,
+    multipartDelete: true,
+    objectDeleteVersion: true,
```
Comment (francoisferrand) [inline lib/policyEvaluator/RequestContext.ts]: not really a quota _check_ in that case : in the sense that these API must never fail due to the quota....
e.g. if quota is way higher than expected (i.e. due to overshoot or because it was lowever), the usage "after" delete may still be higher than quota, but should be accepted nonetheless...

→ we must make sure these 3 API should never return quotaExceeded

Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts]: All delete APIs trigger quota evaluation logic - the logic, in case of quota exceeded, would first increase the inflights (scuba is not aware of the quota) and, upon quota exceeded detection, clear the last increase. In the case of deletes, we simply decrese everytime, and (in cloudserver), we explicitly ensure that any deletion always pass.

As this is something to be done in cloudserver, is this comment about the variable naming?

Comment (francoisferrand) [reply lib/policyEvaluator/RequestContext.ts]: Yes, making sure it cannot fail may indeed be in cloudserver. So:
* variable naming here may be misleading : and could cause confusion (or bugs) when people change (cloudserver) generic code in the future...
* checking the quota "check" may be expensive and we don't need the full evaluation in case of deletes (as you mentioned), only to track the "decrease" : so it would seem more appropriate to differentiate between these case, possibly returning 3 values (do not check, check quota, decrease usage)...

Basically, by bundling all APIs in the same "needQuotasCheck" we introduce coupling between this list of API and the details of the current implementation.... which should be avoided, to facilitate future maintenance.

Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts]: Makes sense, we can split this variable into two: need quota check (increase), need quota check (decrease). Quota exceeded case will anyway require an additional call. In the case of deletes, we only issue a single call anyway.

Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts]: Done in https://github.com/scality/Arsenal/pull/2233/commits/c5ab5d034c5b988c7a5c253427c4ce6237661604

**Comment by ghost** [comment]
@ reviewers; added https://github.com/scality/Arsenal/pull/2233/commits/63cc54d3fa3c859773cabeef840fe8aad9e3ccae to compute the number of bucket with quota as part of the periodic report.

File: lib/policyEvaluator/RequestContext.ts [hunk 2]
```diff
@@ -15,9 +15,19 @@ import {
...
-const _actionNeedQuotaCheck = {
+export const actionNeedQuotaCheck = {
     objectPut: true,
+    objectPutVersion: true,
     objectPutPart: true,
+    objectRestore: true,
+};
+
+// This variable describe APIs that change the bytes
+// stored, requiring quota updates
```
Comment (BourgoisMickael) [inline lib/policyEvaluator/RequestContext.ts]: It can be nice to use jsdoc `/** */` to have description in tooltip on hover in vscode

File: lib/policyEvaluator/RequestContext.ts [hunk 3]
```diff
@@ -15,9 +15,19 @@ import {
...
+export const actionWithDataDeletion = {
```
Comment (BourgoisMickael) [inline lib/policyEvaluator/RequestContext.ts:29]: I think there is also the POST `multiObjectDelete`

File: lib/policyEvaluator/RequestContext.ts [hunk 4]
```diff
@@ -15,9 +15,19 @@ import {
...
+    objectDelete: true,
+    objectDeleteVersion: true,
```
Comment (BourgoisMickael) [inline lib/policyEvaluator/RequestContext.ts:31]: It depends where you take the `apiMethod`, it seems you take it from `request.apiMethod`, in that case, I think there is no `objectDeleteVersion` (https://github.com/scality/Arsenal/blob/development/8.1/lib/s3routes/routes/routeDELETE.ts#L52) it is mixed in the `objectDelete` handler.

The `Version` suffix is added in the RequestContext for Vault https://github.com/scality/cloudserver/blob/development/8.7/lib/api/apiUtils/authorization/prepareRequestContexts.js#L61

File: lib/policyEvaluator/RequestContext.ts (hunk 3 above)
Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts:29]: this API will individually call the deleteObject internal name, so this is not needed

edit: multiObjectDelete does not delete anything, it's the subsequent (internal) calls that are deleting the objects provided as the input, after getting their metadata and authorizing with vault, for each object.

File: lib/policyEvaluator/RequestContext.ts (hunk 4 above)
Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts:31]: When evaluating the quotas, we do it for each of the returned actions from the authorization results. Today it's only the action provided as the input, but we might eventually have more actions (see below). These actions can have `Version` suffix.

We use the `request.apiMethod` only at the end, for the monitoring, where it will indeed not be `objectDeleteVersion`. So we need to support them.

Now, for the multi object delete, indeed I think I am not properly handling it in the internal additional calls, when we authorize one by one all the objects we want to delete. I will work on that.

File: lib/policyEvaluator/RequestContext.ts (hunk 2 above)
Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts]: See https://github.com/scality/Arsenal/pull/2233/commits/c412ada50646f284e9863d5a3f1b0adea951b91e

**Review by BourgoisMickael — APPROVED** [review]

File: lib/policyEvaluator/RequestContext.ts (hunk 4 above)
Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts:31]: Just an update: I added the inflights for the multi delete in cloudserver.

File: lib/policyEvaluator/RequestContext.ts (hunk 3 above)
Comment (ghost) [reply lib/policyEvaluator/RequestContext.ts:29]: Finally added it for another reason, vault uses the _needQuota flag in the request context to output the info about the quota from the account; we thus need this value if we want to be able to evaluate account quotas in this API

**Review by francoisferrand — APPROVED** [review]

File: index.ts [hunk 5]
```diff
@@ -67,6 +69,8 @@ export const policies = {
     RequestContext,
     requestUtils,
     actionMaps,
+    actionNeedQuotaCheck,
+    actionWithDataDeletion,
```
Comment (francoisferrand) [inline index.ts:74]: `actionNeedQuotaRelease` ?

would be better to convey the relation with quota...

**Review by francoisferrand — APPROVED** [review]
